README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [tail-call-proxy](#tail-call-proxy)
  - [Functions](#functions)
    - [lazy](#lazy)
    - [parasitic](#parasitic)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->


<a name="readmemd"></a>

# tail-call-proxy

Delayed initialized objects that support tail-call optimization.

[![npm package][npm-img]][npm-url]
[![Build Status][build-img]][build-url]
[![Downloads][downloads-img]][downloads-url]
[![Issues][issues-img]][issues-url]
[![Code Coverage][codecov-img]][codecov-url]
[![Commitizen Friendly][commitizen-img]][commitizen-url]
[![Semantic Release][semantic-release-img]][semantic-release-url]

[build-img]:https://github.com/Atry/tail-call-proxy/actions/workflows/release.yml/badge.svg
[build-url]:https://github.com/Atry/tail-call-proxy/actions/workflows/release.yml
[downloads-img]:https://img.shields.io/npm/dt/tail-call-proxy
[downloads-url]:https://www.npmtrends.com/tail-call-proxy
[npm-img]:https://img.shields.io/npm/v/tail-call-proxy
[npm-url]:https://www.npmjs.com/package/tail-call-proxy
[issues-img]:https://img.shields.io/github/issues/Atry/tail-call-proxy
[issues-url]:https://github.com/Atry/tail-call-proxy/issues
[codecov-img]:https://codecov.io/gh/Atry/tail-call-proxy/branch/main/graph/badge.svg
[codecov-url]:https://codecov.io/gh/Atry/tail-call-proxy
[semantic-release-img]:https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
[semantic-release-url]:https://github.com/semantic-release/semantic-release
[commitizen-img]:https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
[commitizen-url]:http://commitizen.github.io/cz-cli/

## Functions

### lazy

▸ **lazy**<`T`\>(`tailCall`): `T`

Returns an proxy object backed by `tailCall`, which will be lazily created
at the first time its properties or methods are used.

`lazy` can eliminate tail calls, preventing stack overflow errors in tail
recursive functions or mutual recursive functions.

**`Example`**

The initializer passed to `lazy` should not be called until the first time
`lazyObject.hello` is accessed. When `lazyObject.hello` is accessed more than
once, the second access would not trigger the initializer again.

```typescript doctest
import { lazy } from 'tail-call-proxy';

let counter = 0;
const lazyObject = lazy(() => {
  counter++;
  return { hello: 'world' };
});
expect(counter).toBe(0);

expect(lazyObject.hello).toBe('world');
expect(counter).toBe(1);

expect(lazyObject.hello).toBe('world');
expect(counter).toBe(1);
```

**`Example`**

Note that errors thrown in the initializer will be delayed as well.

```typescript doctest
import { lazy } from 'tail-call-proxy';

let counter = 0;
const lazyError: Record<string, unknown> = lazy(() => {
  counter++;
  throw new Error();
});

// No error is thrown, given that the underlying object have not been created
// yet.
expect(counter).toBe(0);

expect(() => lazyError.toString()).toThrow();
expect(counter).toBe(1);

expect(() => lazyError.toLocaleString()).toThrow();
expect(counter).toBe(1);
```

**`Example`**

The following mutual recursive functions would result in stack overflow:

```typescript doctest
function isEven(n: number): Boolean {
  if (n === 0) {
    return new Boolean(true);
  }
  return isOdd(n - 1);
}

function isOdd(n: number): Boolean {
  if (n === 0) {
    return new Boolean(false);
  }
  return isEven(n - 1);
}

expect(() => isOdd(1000000)).toThrow();
```

However, if you replace `return xxx` with `return lazy(() => xxx)`, it will
use a constant size of stack memory and avoid the stack overflow.

```typescript doctest
import { lazy } from 'tail-call-proxy';
function isEven(n: number): Boolean {
  if (n === 0) {
    return new Boolean(true);
  }
  return lazy(() => isOdd(n - 1));
}

function isOdd(n: number): Boolean {
  if (n === 0) {
    return new Boolean(false);
  }
  return lazy(() => isEven(n - 1));
}

expect(isOdd(1000000).valueOf()).toBe(false);
```

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `tailCall` | () => `T` | the function to create the underlying object |

#### Returns

`T`

#### Defined in

[index.ts:253](https://github.com/Atry/tail-call-proxy/blob/1d1f6c9/src/index.ts#L253)

___

### parasitic

▸ **parasitic**<`T`\>(`tailCall`): `T`

Performs a tail call as soon as possible.

`parasitic` returns either exactly the object returned by `tailCall`, or a
proxy object backed by the object returned by `tailCall`, if there are any
previously started pending tail calls. In the latter case, the underlying
object will be created after all the previous tail calls are finished.

**`Example`**

Unlike [lazy](#lazy), `parasitic` performs the initialization as soon as
possible:

```typescript doctest
import { parasitic } from 'tail-call-proxy';

let counter = 0;
const parasiticObject = parasitic(() => {
  counter++;
  return { hello: 'world' };
});
expect(counter).toBe(1);

expect(parasiticObject.hello).toBe('world');
expect(counter).toBe(1);

expect(parasiticObject.hello).toBe('world');
expect(counter).toBe(1);
```

**`Example`**

`parasitic` is useful when you need tail call optimization while you don't
need the lazy evaluation. It can be used together with [lazy](#lazy)
alternately.

```typescript doctest
import { lazy, parasitic } from 'tail-call-proxy';

let isEvenCounter = 0;
const trueObject = new Boolean(true);
function isEven(n: number): Boolean {
  isEvenCounter++;
  if (n === 0) {
    return trueObject;
  }
  return lazy(() => isOdd(n - 1));
};

let isOddCounter = 0;
const falseObject = new Boolean(false);
function isOdd(n: number): Boolean {
  isOddCounter++;
  if (n === 0) {
    return falseObject;
  }
  return parasitic(() => isEven(n - 1));
};

try {
  // `isEven` is called, but `lazy(() => isOdd(n - 1))` does not trigger
  // `isOdd` immediately.
  const is1000000Even = isEven(1000000);
  expect(isOddCounter).toBe(0);
  expect(isEvenCounter).toBe(1);

  // `valueOf` triggers the rest of the recursion.
  expect(is1000000Even.valueOf()).toBe(true);
  expect(isOddCounter).toBe(500000);
  expect(isEvenCounter).toBe(500001);

  // `is1000000Even` is a lazy proxy backed by `trueObject`, not the exactly
  // same object of `trueObject`.
  expect(is1000000Even).not.toStrictEqual(trueObject);
  expect(is1000000Even).toEqual(trueObject);
} finally {
  isEvenCounter = 0;
  isOddCounter = 0;
}

// `isOdd` is called, in which `parasitic(() => isEven(n - 1))` triggers the
// rest of the recursion immediately.
const is1000000Odd = isOdd(1000000);
expect(isOddCounter).toBe(500001);
expect(isEvenCounter).toBe(500000);
expect(is1000000Odd.valueOf()).toBe(false);
expect(isOddCounter).toBe(500001);
expect(isEvenCounter).toBe(500000);

// `is1000000Odd` is exactly the same object of `falseObject`, not a lazy
// proxy.
expect(is1000000Odd).toStrictEqual(falseObject);
```

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `tailCall` | () => `T` | the function to create the underlying object |

#### Returns

`T`

#### Defined in

[index.ts:353](https://github.com/Atry/tail-call-proxy/blob/1d1f6c9/src/index.ts#L353)


CHANGELOG.md
## [1.1.12](https://github.com/Atry/tail-call-proxy/compare/v1.1.11...v1.1.12) (2023-01-29)

## [1.1.11](https://github.com/Atry/tail-call-proxy/compare/v1.1.10...v1.1.11) (2023-01-26)

## [1.1.10](https://github.com/Atry/tail-call-proxy/compare/v1.1.9...v1.1.10) (2023-01-26)


### Dependencies and Other Build Updates

* add a keyword ([9707415](https://github.com/Atry/tail-call-proxy/commit/9707415925277f4eee68ff1c52e7769a5087d3c3))

## [1.1.9](https://github.com/Atry/tail-call-proxy/compare/v1.1.8...v1.1.9) (2023-01-16)


### Dependencies and Other Build Updates

* **deps:** remove unused doctest-ts-improved ([7067d45](https://github.com/Atry/tail-call-proxy/commit/7067d458f33acc2459d65042133ac6cd7668db39))

## [1.1.8](https://github.com/Atry/tail-call-proxy/compare/v1.1.7...v1.1.8) (2023-01-16)


### Dependencies and Other Build Updates

* **deps:** update prettier ([e97b441](https://github.com/Atry/tail-call-proxy/commit/e97b44186a413bb3b2b2241300bd28de7284e2d0))

## [1.1.7](https://github.com/Atry/tail-call-proxy/compare/v1.1.6...v1.1.7) (2023-01-16)


### Bug Fixes

* remove unused imports ([843317f](https://github.com/Atry/tail-call-proxy/commit/843317f8f7dacbe3d107c605094d6e7017af4e9e))

## [1.1.6](https://github.com/Atry/tail-call-proxy/compare/v1.1.5...v1.1.6) (2023-01-16)


### Bug Fixes

* a type checking error ([e0aaeaf](https://github.com/Atry/tail-call-proxy/commit/e0aaeaf93499940a2751cfa209bd84f5309fce50))
* use getPrototypeOf to implement toStringTag ([83a1d41](https://github.com/Atry/tail-call-proxy/commit/83a1d411ed71599cc7f69616193d9d06daaaf952))

## [1.1.5](https://github.com/Atry/tail-call-proxy/compare/v1.1.4...v1.1.5) (2023-01-14)


### Bug Fixes

* let the proxy mimic more behavior ([1a96c0a](https://github.com/Atry/tail-call-proxy/commit/1a96c0ab38f80449238da390058b32e84ec287d5))


### Dependencies and Other Build Updates

* add moduleNameMapper ([a09b0d2](https://github.com/Atry/tail-call-proxy/commit/a09b0d287cb8cc9408e8a5c4fda475527ff3cbd6))
* add modulePaths ([6be3cad](https://github.com/Atry/tail-call-proxy/commit/6be3cadd6ef2d962f0d8b886dcf759a638f93292))
* add paths mapping ([a79b776](https://github.com/Atry/tail-call-proxy/commit/a79b7767df5ffb9d4bb1291d9621c4f521682ee0))
* add roots setting ([031a119](https://github.com/Atry/tail-call-proxy/commit/031a119c2a8c25b6a0485c2fff90297564f45c0a))
* remove unused settings ([#9](https://github.com/Atry/tail-call-proxy/issues/9)) ([eb538b0](https://github.com/Atry/tail-call-proxy/commit/eb538b081eba9545aa299c5315c39b0af676d574))

## [1.1.4](https://github.com/Atry/tail-call-proxy/compare/v1.1.3...v1.1.4) (2023-01-14)

## [1.1.3](https://github.com/Atry/tail-call-proxy/compare/v1.1.2...v1.1.3) (2023-01-14)

## [1.1.2](https://github.com/Atry/tail-call-proxy/compare/v1.1.1...v1.1.2) (2023-01-14)

## [1.1.1](https://github.com/Atry/tail-call-proxy/compare/v1.1.0...v1.1.1) (2023-01-14)

## [1.1.0](https://github.com/Atry/tail-call-proxy/compare/v1.0.10...v1.1.0) (2023-01-14)


### Features

* **index:** change to bump minor version ([#7](https://github.com/Atry/tail-call-proxy/issues/7)) ([84cb8c4](https://github.com/Atry/tail-call-proxy/commit/84cb8c44d0dd5143ecd1d12b129b426283d10ef5))
* **template:** initial commit ([e37f2fb](https://github.com/Atry/tail-call-proxy/commit/e37f2fbb4998ac87a8c00bf032b30163d2e809d2))


### Bug Fixes

* **cve:** upgrade semantic-release to fix GHSA-rrrm-qjm4-v8hf ([d9856cd](https://github.com/Atry/tail-call-proxy/commit/d9856cd7ce303bad4dbac29c29ee01d97cd4b9fa))
* **deps:** cve-2021-23425 ([ecc8553](https://github.com/Atry/tail-call-proxy/commit/ecc8553254d6724195472dd644f2e83df16cb93c))
* **deps:** cve-2021-23440 ([8c88925](https://github.com/Atry/tail-call-proxy/commit/8c889253bf6a9a5209816b13398b7f7f1cdae416))
* **deps:** update caniuse db for jest deprecation warning ([29eb3cc](https://github.com/Atry/tail-call-proxy/commit/29eb3ccdf8e64243095f78a953e1498b5487f044))
* exclude test files from build ([#19](https://github.com/Atry/tail-call-proxy/issues/19)) ([00517e8](https://github.com/Atry/tail-call-proxy/commit/00517e8aaa1b4d0754e7f99db137c7b817672e08))
* **license:** update copyright date in license ([e83518d](https://github.com/Atry/tail-call-proxy/commit/e83518d3dd5f0313013bacbe8fb9c68932d54945))


### Dependencies and Other Build Updates

* **deps-dev:** bump @types/jest to 27 ([3331b8d](https://github.com/Atry/tail-call-proxy/commit/3331b8d4b007d8a4bd4ea0fd33de8b37b7fdda29))
* **deps-dev:** bump semantic-release from 19.0.2 to 19.0.3 ([#15](https://github.com/Atry/tail-call-proxy/issues/15)) ([89b8174](https://github.com/Atry/tail-call-proxy/commit/89b8174ecf7d951f57f72579f1ba18e627d132f4))
* **deps:** bump json5 from 2.2.0 to 2.2.3 ([#24](https://github.com/Atry/tail-call-proxy/issues/24)) ([5bcadc3](https://github.com/Atry/tail-call-proxy/commit/5bcadc3aa3cdcf8323622a0fad3aa9f33ad9bf88))
* **deps:** bump normalize-url from 6.0.0 to 6.0.1 ([#4](https://github.com/Atry/tail-call-proxy/issues/4)) ([8166b9a](https://github.com/Atry/tail-call-proxy/commit/8166b9ace81832579218ad88382e400fc9a1bdc3))
* **deps:** bump npm from 8.4.0 to 8.12.0 ([#13](https://github.com/Atry/tail-call-proxy/issues/13)) ([d394432](https://github.com/Atry/tail-call-proxy/commit/d39443238110052106418d318bee74430d9890bc))
* **deps:** bump path-parse from 1.0.6 to 1.0.7 ([#6](https://github.com/Atry/tail-call-proxy/issues/6)) ([9cea74b](https://github.com/Atry/tail-call-proxy/commit/9cea74b30748fd73b2865812d50c2cf460e24ea8))
* **deps:** bump semver-regex from 3.1.3 to 3.1.4 ([#14](https://github.com/Atry/tail-call-proxy/issues/14)) ([d1c7b3e](https://github.com/Atry/tail-call-proxy/commit/d1c7b3e29ab425de319382a81a594d73e2e51896))
* **deps:** bump trim-newlines from 3.0.0 to 3.0.1 ([#5](https://github.com/Atry/tail-call-proxy/issues/5)) ([2e22195](https://github.com/Atry/tail-call-proxy/commit/2e22195d68efd6c5b69e44bb6be5a6c0f66e4237))

test/index.spec.ts
import { parasitic, lazy } from '../src';

describe('index', () => {
  describe('parasitic', () => {
    it('should actually return a number from a number recursive function', () => {
      // eslint-disable-next-line @typescript-eslint/ban-types
      function go(i: number): Number {
        if (i < 1000000) {
          return parasitic(() => go(i + 1));
        } else {
          return new Number(42);
        }
      }
      expect(go(0).valueOf()).toBe(42);
    });

    it('should actually return numbers from nested number recursive functions', () => {
      function fortyTwo(i: number): { x: number } {
        if (i < 1000000) {
          return parasitic(() => fortyTwo(i + 1));
        } else {
          return { x: 42 };
        }
      }
      type List =
        | { value: { x: number }; next: List }
        | { value?: undefined; next?: undefined };
      function list(i: number, accumulator: List): List {
        if (i < 3) {
          return parasitic(() =>
            list(i + 1, { value: fortyTwo(0), next: accumulator })
          );
        } else {
          return accumulator;
        }
      }
      expect(list(0, {})?.value?.x).toBe(42);
    });
  });
  describe('lazy', () => {
    it('should actually return a number from a number recursive function', () => {
      // eslint-disable-next-line @typescript-eslint/ban-types
      function go(i: number): Number {
        if (i < 1000000) {
          return lazy(() => go(i + 1));
        } else {
          return new Number(42);
        }
      }
      expect(go(0).valueOf()).toBe(42);
    });
    it('should not throw a stack overflow error from an asynchronous recursive functions', async () => {
      async function go(i: number): Promise<number> {
        if (i < 3) {
          return lazy(() => go(i + 1));
        } else {
          return 42;
        }
      }
      expect(await go(0)).toBe(42);
    });
    it('should execute aliases of a tail call proxy only once', () => {
      let count = 0;
      const buffer: { x: number }[] = [];
      function fortyTwo(i: number): { x: number } {
        if (i < 3) {
          return lazy(() => fortyTwo(i + 1));
        } else {
          count++;
          return { x: 42 };
        }
      }
      function sideEffect(): { x: number } {
        const result = fortyTwo(0);
        buffer.push(result);
        return result;
      }
      function tailCallSideEffect(): { x: number } {
        return lazy(() => sideEffect());
      }
      type List =
        | { value: { x: number }; next: List }
        | { value?: undefined; next?: undefined };
      function list(i: number, accumulator: List): List {
        if (i < 2) {
          return lazy(() =>
            list(i + 1, { value: tailCallSideEffect(), next: accumulator })
          );
        } else {
          return accumulator;
        }
      }
      expect(list(0, {})).toMatchObject({
        value: { x: 42 },
        next: { value: { x: 42 }, next: {} },
      });
      expect(buffer).toMatchObject([{ x: 42 }, { x: 42 }]);
      expect(count).toBe(2);
    });
  });
});


src/index.ts
/* eslint-disable @typescript-eslint/ban-types */

/**
 * Delayed initialized objects that support tail-call optimization.
 *
 * [![npm package][npm-img]][npm-url]
 * [![Build Status][build-img]][build-url]
 * [![Downloads][downloads-img]][downloads-url]
 * [![Issues][issues-img]][issues-url]
 * [![Code Coverage][codecov-img]][codecov-url]
 * [![Commitizen Friendly][commitizen-img]][commitizen-url]
 * [![Semantic Release][semantic-release-img]][semantic-release-url]
 *
 * [build-img]:https://github.com/Atry/tail-call-proxy/actions/workflows/release.yml/badge.svg
 * [build-url]:https://github.com/Atry/tail-call-proxy/actions/workflows/release.yml
 * [downloads-img]:https://img.shields.io/npm/dt/tail-call-proxy
 * [downloads-url]:https://www.npmtrends.com/tail-call-proxy
 * [npm-img]:https://img.shields.io/npm/v/tail-call-proxy
 * [npm-url]:https://www.npmjs.com/package/tail-call-proxy
 * [issues-img]:https://img.shields.io/github/issues/Atry/tail-call-proxy
 * [issues-url]:https://github.com/Atry/tail-call-proxy/issues
 * [codecov-img]:https://codecov.io/gh/Atry/tail-call-proxy/branch/main/graph/badge.svg
 * [codecov-url]:https://codecov.io/gh/Atry/tail-call-proxy
 * [semantic-release-img]:https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
 * [semantic-release-url]:https://github.com/semantic-release/semantic-release
 * [commitizen-img]:https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
 * [commitizen-url]:http://commitizen.github.io/cz-cli/
 *
 * @module
 */

import {
  getTarget,
  AllowGetTarget,
  Mapped,
  DefaultToPrimitive,
  DefaultToStringTag,
  TargetAsThis,
} from 'proxy-handler-decorators';
import { DefaultProxyHandler } from 'default-proxy-handler';

type ProxyTarget<T extends object> = {
  tailCall?: () => T;
  tail?: ProxyTarget<T>;
  result?: T;
  error?: any;
};

function getResult<T extends object>(target: ProxyTarget<T>): T {
  if ('result' in target) {
    return target.result as T;
  } else if ('error' in target) {
    throw target.error;
  } else {
    run(target);
    return getResult(target);
  }
}

@AllowGetTarget
// eslint-disable-next-line @typescript-eslint/no-unsafe-call
@Mapped(getResult)
@DefaultToPrimitive
@DefaultToStringTag
@TargetAsThis
class TailCallProxyHandler<T extends object> extends DefaultProxyHandler<
  T | ProxyTarget<T>
> {}

let parasiticQueue: ProxyTarget<any>[] = [];

let isRunning = false;

function step<T extends object>(target: ProxyTarget<T>): void {
  const { tailCall } = target;
  let proxyOrResult: T;
  if (tailCall !== undefined) {
    try {
      try {
        proxyOrResult = tailCall();
      } finally {
        delete target.tailCall;
      }
    } catch (error) {
      target.error = error;
      return;
    }
    if (proxyOrResult == null) {
      target.result = proxyOrResult;
    } else {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const tail: ProxyTarget<T> | undefined = getTarget(proxyOrResult);
      if (tail === undefined) {
        target.result = proxyOrResult;
      } else {
        target.tail = tail;
      }
    }
  }
}

function resolveWith<T extends object>(
  headTarget: ProxyTarget<T>,
  lastTarget: ProxyTarget<T>
): void {
  if ('error' in lastTarget) {
    headTarget.error = lastTarget.error;
  } else {
    headTarget.result = lastTarget.result;
  }
}

function resolve<T extends object>(headTarget: ProxyTarget<T>): void {
  step(headTarget);
  if (headTarget.tail !== undefined) {
    let target = headTarget.tail;
    delete headTarget.tail;
    for (;;) {
      step(target);
      const { tail } = target;
      const queue = parasiticQueue;
      if (queue.length !== 0) {
        parasiticQueue = [];
        for (const parasiticTarget of queue) {
          if (parasiticTarget !== tail) {
            resolve(parasiticTarget);
          }
        }
      }
      if (tail !== undefined) {
        target = tail;
      } else {
        resolveWith(headTarget, target);
        return;
      }
    }
  }
}

function run<T extends object>(target: ProxyTarget<T>): void {
  if (isRunning) {
    resolve(target);
  }
  isRunning = true;
  try {
    resolve(target);
  } finally {
    isRunning = false;
  }
}

const LAZY_PROXY_HANDLER = new TailCallProxyHandler();
/**
 * Returns an proxy object backed by `tailCall`, which will be lazily created
 * at the first time its properties or methods are used.
 *
 * `lazy` can eliminate tail calls, preventing stack overflow errors in tail
 * recursive functions or mutual recursive functions.
 *
 * @param tailCall the function to create the underlying object
 *
 * @example
 *
 * The initializer passed to `lazy` should not be called until the first time
 * `lazyObject.hello` is accessed. When `lazyObject.hello` is accessed more than
 * once, the second access would not trigger the initializer again.
 *
 * ```typescript doctest
 * import { lazy } from 'tail-call-proxy';
 *
 * let counter = 0;
 * const lazyObject = lazy(() => {
 *   counter++;
 *   return { hello: 'world' };
 * });
 * expect(counter).toBe(0);
 *
 * expect(lazyObject.hello).toBe('world');
 * expect(counter).toBe(1);
 *
 * expect(lazyObject.hello).toBe('world');
 * expect(counter).toBe(1);
 * ```
 *
 * @example
 *
 * Note that errors thrown in the initializer will be delayed as well.
 *
 * ```typescript doctest
 * import { lazy } from 'tail-call-proxy';
 *
 * let counter = 0;
 * const lazyError: Record<string, unknown> = lazy(() => {
 *   counter++;
 *   throw new Error();
 * });
 *
 * // No error is thrown, given that the underlying object have not been created
 * // yet.
 * expect(counter).toBe(0);
 *
 * expect(() => lazyError.toString()).toThrow();
 * expect(counter).toBe(1);
 *
 * expect(() => lazyError.toLocaleString()).toThrow();
 * expect(counter).toBe(1);
 * ```
 *
 * @example
 *
 * The following mutual recursive functions would result in stack overflow:
 *
 * ```typescript doctest
 * function isEven(n: number): Boolean {
 *   if (n === 0) {
 *     return new Boolean(true);
 *   }
 *   return isOdd(n - 1);
 * }
 *
 * function isOdd(n: number): Boolean {
 *   if (n === 0) {
 *     return new Boolean(false);
 *   }
 *   return isEven(n - 1);
 * }
 *
 * expect(() => isOdd(1000000)).toThrow();
 * ```
 *
 * However, if you replace `return xxx` with `return lazy(() => xxx)`, it will
 * use a constant size of stack memory and avoid the stack overflow.
 *
 * ```typescript doctest
 * import { lazy } from 'tail-call-proxy';
 * function isEven(n: number): Boolean {
 *   if (n === 0) {
 *     return new Boolean(true);
 *   }
 *   return lazy(() => isOdd(n - 1));
 * }
 *
 * function isOdd(n: number): Boolean {
 *   if (n === 0) {
 *     return new Boolean(false);
 *   }
 *   return lazy(() => isEven(n - 1));
 * }
 *
 * expect(isOdd(1000000).valueOf()).toBe(false);
 * ```
 */
export function lazy<T extends object>(tailCall: () => T): T {
  return new Proxy({ tailCall }, LAZY_PROXY_HANDLER) as T;
}

/**
 * Performs a tail call as soon as possible.
 *
 * `parasitic` returns either exactly the object returned by `tailCall`, or a
 * proxy object backed by the object returned by `tailCall`, if there are any
 * previously started pending tail calls. In the latter case, the underlying
 * object will be created after all the previous tail calls are finished.
 *
 * @param tailCall the function to create the underlying object
 *
 * @example
 *
 * Unlike {@link lazy}, `parasitic` performs the initialization as soon as
 * possible:
 *
 * ```typescript doctest
 * import { parasitic } from 'tail-call-proxy';
 *
 * let counter = 0;
 * const parasiticObject = parasitic(() => {
 *   counter++;
 *   return { hello: 'world' };
 * });
 * expect(counter).toBe(1);
 *
 * expect(parasiticObject.hello).toBe('world');
 * expect(counter).toBe(1);
 *
 * expect(parasiticObject.hello).toBe('world');
 * expect(counter).toBe(1);
 * ```
 *
 * @example
 *
 * `parasitic` is useful when you need tail call optimization while you don't
 * need the lazy evaluation. It can be used together with {@link lazy}
 * alternately.
 *
 * ```typescript doctest
 * import { lazy, parasitic } from 'tail-call-proxy';
 *
 * let isEvenCounter = 0;
 * const trueObject = new Boolean(true);
 * function isEven(n: number): Boolean {
 *   isEvenCounter++;
 *   if (n === 0) {
 *     return trueObject;
 *   }
 *   return lazy(() => isOdd(n - 1));
 * };
 *
 * let isOddCounter = 0;
 * const falseObject = new Boolean(false);
 * function isOdd(n: number): Boolean {
 *   isOddCounter++;
 *   if (n === 0) {
 *     return falseObject;
 *   }
 *   return parasitic(() => isEven(n - 1));
 * };
 *
 * try {
 *   // `isEven` is called, but `lazy(() => isOdd(n - 1))` does not trigger
 *   // `isOdd` immediately.
 *   const is1000000Even = isEven(1000000);
 *   expect(isOddCounter).toBe(0);
 *   expect(isEvenCounter).toBe(1);
 *
 *   // `valueOf` triggers the rest of the recursion.
 *   expect(is1000000Even.valueOf()).toBe(true);
 *   expect(isOddCounter).toBe(500000);
 *   expect(isEvenCounter).toBe(500001);
 *
 *   // `is1000000Even` is a lazy proxy backed by `trueObject`, not the exactly
 *   // same object of `trueObject`.
 *   expect(is1000000Even).not.toStrictEqual(trueObject);
 *   expect(is1000000Even).toEqual(trueObject);
 * } finally {
 *   isEvenCounter = 0;
 *   isOddCounter = 0;
 * }
 *
 * // `isOdd` is called, in which `parasitic(() => isEven(n - 1))` triggers the
 * // rest of the recursion immediately.
 * const is1000000Odd = isOdd(1000000);
 * expect(isOddCounter).toBe(500001);
 * expect(isEvenCounter).toBe(500000);
 * expect(is1000000Odd.valueOf()).toBe(false);
 * expect(isOddCounter).toBe(500001);
 * expect(isEvenCounter).toBe(500000);
 *
 * // `is1000000Odd` is exactly the same object of `falseObject`, not a lazy
 * // proxy.
 * expect(is1000000Odd).toStrictEqual(falseObject);
 * ```
 */
export function parasitic<T extends object>(tailCall: () => T): T {
  const target: ProxyTarget<T> = { tailCall };
  if (isRunning) {
    parasiticQueue.push(target);
    return new Proxy(target, LAZY_PROXY_HANDLER) as T;
  } else {
    return getResult(target);
  }
}
